asterius/test/ghc-testsuite.hs

```diff
@@ -37,27 +64,169 @@ getTestCases = do
         readFileNullable (c -<.> "stdout") <*>
         readFileNullable (c -<.> "stderr")
 
-runTestCase :: TestCase -> IO (LBS.ByteString, LBS.ByteString)
+
+
+
+data TestOutcome = TestSuccess | TestFailure deriving(Eq, Show, Generic)
+
+instance ToJSON TestOutcome where
+
+
+data TestRecord = TestRecord
+  { trOutcome :: !TestOutcome
+  , trPath :: !FilePath -- ^ Path of the test case
+  , trErrorMessage :: !String -- ^ If the test failed, then the error message associated to the failure.
+  } deriving(Generic)
+
+instance ToJSON TestRecord where
+
+
+
+
+-- | Log of tests that have run
+newtype TestLog = TestLog [TestRecord] deriving(Semigroup, Monoid, Generic)
+
+instance ToJSON TestLog where
+
+-- | Append a value to the test log in safe way when we have multiple threads
+consTestLog :: TestRecord -> IORef TestLog -> IO ()
+consTestLog tr tlref = modifyIORef' tlref (\(TestLog tl) -> TestLog $ tr:tl)
+
+
+-- | What happened when you tried to run the test
+data RunOutcome = RunSuccess | RunFailure String deriving(Eq)
+
+
+-- [Note: RunOutcome's Show instance]
+-- | Have the Show instance print the exception after the separator  so we can
+-- | strip out the separator in the printer
+-- | This way, our custom ingredient can still serialize all the information
+-- | that comes after the ``, but when we print, we strip out the leading `:`
+-- | and all text that follows it.
+separator :: Char
+separator = 'γ'
+
+instance Show RunOutcome where
+  show (RunSuccess) = "RunSuccess"
+  show (RunFailure e) = "RunFailure" <> [separator] <> (show e)
+
+instance ToJSON RunOutcome where
+    toJSON RunSuccess = toJSON . show $ RunSuccess
+    toJSON (RunFailure f) = toJSON $ "RunFailure(" <> show f <> ")"
+
+runTestCase :: TestCase -> IO ()
 runTestCase TestCase {..} = do
-  _ <- readProcess "ahc-link" ["--input-hs", casePath] ""
+  _ <- readProcess "ahc-link" ["--input-hs", casePath, "--binaryen"] ""
   mod_buf <- LBS.readFile $ casePath -<.> "wasm"
   withJSSession defJSSessionOpts $ \s -> do
     i <- newAsteriusInstance s (casePath -<.> "lib.mjs") mod_buf
     hsInit s i
-    hsMain s i
+    -- | Try to run main. If we throw an exception, return a
```

Comment:
It's good to consider the case where the exported `main()` might fail here, and capture it in `RunFailure`. It would be nice to make the error handler here more coarse-grained, to also consider that even `hs_init()`, `newAsteriusInstance`, any logic within the `withJSSession` block might also fail, albeit much less likely than `main()` failures. Moving the `catch` a level up to handle the whole `withJSSession` call is somewhat safer.

And even the logic before the `withJSSession` block is entered might throw. So it might make sense to also add a `CompileFailure e` variant to `RunOutcome`, and use another `catch` to handle it.

Reply:
Right, this makes sense. I'll change the test case accordingly.

Reply:
Not sure if the way in which I implemented this is optimal. We now have a `CompileOutcome` and `RunOutcome`.

I like this because it feels conceptually cleaner. Thoughts?

Reply:
This now looks good to me, thanks for the work.